internal/plugin/health_monitor.go

```diff
@@ -0,0 +1,132 @@
+	"github.com/kubeshop/botkube/pkg/api/source"
+	"github.com/kubeshop/botkube/pkg/config"
+)
+
+type HealthMonitor struct {
+	log                    logrus.FieldLogger
+	logConfig              config.Logger
+	sourceSupervisorChan   chan pluginMetadata
+	executorSupervisorChan chan pluginMetadata
+	schedulerChan          chan string
+	executorsStore         *store[executor.Executor]
+	sourcesStore           *store[source.Source]
+	policy                 config.PluginRestartPolicy
+	pluginRestartStats     map[string]int
+	healthCheckInterval    time.Duration
+}
+
+func NewHealthMonitor(logger logrus.FieldLogger, logCfg config.Logger, policy config.PluginRestartPolicy, schedulerChan chan string, sourceSupervisorChan, executorSupervisorChan chan pluginMetadata, executorsStore *store[executor.Executor], sourcesStore *store[source.Source], healthCheckInterval time.Duration) *HealthMonitor {
+	return &HealthMonitor{
+		log:                    logger,
+		logConfig:              logCfg,
+		policy:                 policy,
+		schedulerChan:          schedulerChan,
+		sourceSupervisorChan:   sourceSupervisorChan,
+		executorSupervisorChan: executorSupervisorChan,
+		executorsStore:         executorsStore,
+		sourcesStore:           sourcesStore,
+		pluginRestartStats:     make(map[string]int),
+		healthCheckInterval:    healthCheckInterval,
+	}
+}
+
+func (m *HealthMonitor) Start(ctx context.Context) {
+	go m.monitorSourcePluginHealth(ctx)
+	go m.monitorExecutorPluginHealth(ctx)
+}
+
+func (m *HealthMonitor) monitorSourcePluginHealth(ctx context.Context) {
+	m.log.Info("Starting source plugin supervisor...")
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case plugin := <-m.sourceSupervisorChan:
+			m.log.Infof("Restarting source plugin %q...", plugin.name)
+			if source, ok := m.sourcesStore.EnabledPlugins.Get(plugin.name); ok && source.Cleanup != nil {
+				m.log.Debugf("Releasing resources of source plugin %q...", plugin.name)
+				source.Cleanup()
+			}
+
+			// botkube/kubernetes
+			repoPluginPair := fmt.Sprintf("%s/%s", plugin.repo, plugin.name)
+			m.sourcesStore.EnabledPlugins.Delete(repoPluginPair)
+
+			if ok := m.shouldRestartPlugin(repoPluginPair); !ok {
+				m.log.Warnf("Plugin %q has been restarted too many times. Deactivating...", plugin.name)
+				continue
+			}
+
+			p, err := createGRPCClient[source.Source](ctx, m.log, m.logConfig, plugin, TypeSource, m.sourceSupervisorChan, m.healthCheckInterval)
+			if err != nil {
+				m.log.WithError(err).Errorf("Failed to restart plugin %q.", plugin.name)
+				continue
+			}
+
+			m.sourcesStore.EnabledPlugins.Insert(repoPluginPair, p)
+			m.schedulerChan <- repoPluginPair
+		}
+	}
+}
+
+func (m *HealthMonitor) monitorExecutorPluginHealth(ctx context.Context) {
+	m.log.Info("Starting executor plugin supervisor...")
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case plugin := <-m.executorSupervisorChan:
+			m.log.Infof("Restarting executor plugin %q...", plugin.name)
```

Comment:
would be nice to print the "status" with the number of retries and max retries like `(attempt no 2 of max 10) `